docs/specification.md

```diff
@@ -357,6 +463,158 @@ unit
 
 ## Proposal entrypoints
 
-**TODO**
+### **propose**
+
+```haskell
+data Proposal proposalMetadata = Proposal
+    { frozen_token_requirement :: Natural
+    -- ^ How many tokens should be frozen in order to vote on this proposal.
+    , metadata :: proposalMetadata
+    }
+
+data Parameter proposalMetadata
+  = Propose Natural (Proposal proposalMetadata)
+```
+
+Parameter (in Michelson):
+```
+(pair %propose
+  nat
+  (pair
+    (nat %frozen_token_requirement)
+    (<proposal_type> %metadata)
+  )
+)
+```
+
+- The proposal is saved and assigned a unique numeric proposal ID.
+- The `Natural` value (let's say `n`) determines how many sender's tokens will be frozen.
+- Sender MUST have enough unfrozen tokens (i. e. `≥ n`).
+- Fails with `PROPOSAL_INSUFFICIENT_BALANCE` if the unfrozen token balance of the SENDER
+  is less than `n`.
+- Fails if the proposal is rejected by `cProposalCheck` from the configuration.
+- The sender's balance in frozen tokens is increased by `n` and in unfrozen tokens is decreased by `n`.
+
+### **proposal_metadata**
+
+```haskell
+type ProposalId = Natural
+
+data Parameter proposalMetadata
+  = ProposalMetadata (View ProposalId (Proposal proposalMetadata))
+```
+
+Parameter (in Michelson):
+```
+(pair %proposal_meatadata
+  (nat %requests)
+  (contract %callback <proposal_type>)
+)
+```
+
+- Return the proposal associated with the id.
+
+- Fails with `PROPOSAL_NOT_EXIST` if the proposal id is
+  not associated with any ongoing proposals.
+
+### **set_voting_period**
+
+```haskell
+data Parameter proposalMetadata
+  = SetVotingPeriod Natural
+```
+
+Parameter (in Michelson):
+```
+nat
+```
+
+- Update how long the voting period lasts for new proposals
+- Fails with `NOT_ADMINISTRATOR` if the sender is not the administrator.
+
+### **set_quorum_threshold**
+
+```haskell
+data Parameter proposalMetadata
+  = SetQuorumThreshold Natural
+```
+
+Parameter (in Michelson):
+```
+nat
+```
+
+- Update the amount of votes which is required for a successful proposal for new proposals
+- Fails with `NOT_ADMINISTRATOR` if the sender is not the administrator.
+
+### **vote**
+
+```haskell
+type ProposalId = Natural
+
+type VoteType = Upvote | Downvote
+
+data VoteParam = VoteParam
+  { proposal_id :: ProposalId
+  , vote_type :: VoteType
+  }
+
+data Parameter proposalMetadata
+  = Vote VoteParam
+```
+
+Parameter (in Michelson):
+
+```
+(pair %vote
+  (nat %proposal_id)
+  (bool %vote_type)
+)
+```
+
+- The `Natural` value (let's say `n`) denotes the amount of tokens required by the proposal to vote.
```

Comment:
The only `Natural` value here is `ProposalId` and it's not the amount of tokens, probably this statement was wrongly copied here.

Reply:
I was meant to introduce a variable `n` as the amount require by the proposal to vote. Would this wording a bit better, Ivan?

`- Let `n` denotes the amount of tokens required by the associated proposal to vote.`

Reply:
Yes, looks good (if you `s/denotes/denote`).

Reply:
Fixed